source/extensions/certificate_providers/certificate_provider.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <list>
+
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+#include "openssl/ssl.h"
+#include "openssl/x509v3.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace CertificateProviders {
+
+struct Certpair {
+  const std::string& certificate_;
+  const std::string& private_key_;
+};
+
+struct Capabilites {
+  /* whether or not a provider provides a ca cert directly */
+  bool provide_ca_cert = false;
+
+  /* Whether or not a provider provides identity certs directly */
+  bool provide_identity_certs = false;
+
+  /* whether or not a provider supports generating identity certs during handshake,
+   * which requires the capability provide_ca_cert to sign the certificates
+   */
+  bool generate_identity_certs = false;
+};
+
+class CertificateProvider {
+public:
+  virtual ~CertificateProvider() = default;
+
+  virtual Capabilites capabilities() const PURE;
+
+  /**
+   * @return CA certificate from provider
+   */
+  virtual const std::string& getCACertificate() const PURE;
+
+  /*
+   * Get TLS identity certificates directly from provider
+   */
+  virtual std::list<Certpair> getTlsCertificates() PURE;
+
+  /*
+   * Generate TLS identity certificate dynamically during TLS handshake
+   */
+  virtual Certpair* generateTlsCertificate(absl::string_view server_name) PURE;
```

Comment:
Static certs are prepared in advance, so we just look up all certs and pick the correct one.
Dynamic certs are generated during handshake, it needs read CLIENT_HELLO to rget at least server name to generate certs, so I'm afraid we can't use the same interface? Do you have some suggestions?

Reply:
Hmm, that's an interesting use-case.

@zhenlian, how would we handle something like this in the gRPC cert provider interface?

Reply:
gRPC supports basic SNI(Server Name Indication): on the server side, we allow users to specify a list of certificates, and gRPC will automatically choose the certificate matching the "server name" info sent from the client side, and send that certificate back to the client. [The SNI implementation](https://github.com/grpc/grpc/blob/master/src/core/tsi/ssl_transport_security.cc#L2292) is shared by both the old ssl credential and the new tls credential.

I said it was "basic", because the logic is deeply inside gRPC TSI layer, and not exposed to users. Sometimes users may want to do some advanced operations, e.g. use the "server name" to consult some third-party service, and then use the information got from the third-party service to decide which certificate to return. gRPC currently doesn't support that. We have a similar use case, and I am in the middle of designing a security API that could allow these cases here, so feel free to let me know if this is also something we should take care of, and I will take that into consideration. Thanks!

Reply:
> gRPC supports basic SNI(Server Name Indication): on the server side, we allow users to specify a list of certificates, and gRPC will automatically choose the certificate matching the "server name" info sent from the client side, and send that certificate back to the client. [The SNI implementation](https://github.com/grpc/grpc/blob/master/src/core/tsi/ssl_transport_security.cc#L2292) is shared by both the old ssl credential and the new tls credential.

"choose" still means these certificates are generated in advance, right? In our proposal https://github.com/envoyproxy/envoy/issues/18928, we clarify the use case, we need to "generate" certs but not "choose" from a cert list.

Reply:
Yes, gRPC SNI just extends the flexibility a bit, from one pre-defined cert chain to a list of pre-defined cert chains. But they are still produced "in advance".
I guess your description seem to match the second case here, we we want to allow users to dynamically generate certificates on the fly, based on the handshake information. gRPC currently doesn't support that.

Reply:
After thinking about this for a while, I think I agree that dynamically generated certs will need a separate API.  We would definitely not be able to support them via the watch-based API that I've argued for above (and which I still think we need for static certs).

This raises the question of whether it makes sense to have a single CertificateProvider API here that provides both static and dynamic certs, or whether we say that CertificateProvider is only for static certs and we need to build some other API for dynamic certs.  I guess either one can work, but it seems to me that it will be unlikely for a given cert provider implementation to do both, so it's not clear what benefit we get from putting them both in the same API.  Therefore, I think I would prefer to see them in a separate API.

Also, regardless of whether we handle dynamic certs as part of CertificateProvider or we add a new API for them, I think we should probably have an async API to handle this case.  Otherwise, we'd be unable to handle cases where generating the cert itself requires some async I/O, such as calling out to an external CA using a gRPC API.

Reply:
we don't have generateTlsCertificate for now.